src/runtime/micro/micro_session.cc

```diff
@@ -0,0 +1,306 @@
+    LOG(FATAL) << "error during micro function execution:\n"
+               << "  " << last_error_str;
+  }
+}
+
+BinaryInfo MicroSession::LoadBinary(std::string binary_path) {
+  SectionLocation text;
+  SectionLocation rodata;
+  SectionLocation data;
+  SectionLocation bss;
+
+  text.size = GetSectionSize(binary_path, kText);
+  rodata.size = GetSectionSize(binary_path, kRodata);
+  data.size = GetSectionSize(binary_path, kData);
+  bss.size = GetSectionSize(binary_path, kBss);
+
+  text.start = AllocateInSection(kText, text.size);
+  rodata.start = AllocateInSection(kRodata, rodata.size);
+  data.start = AllocateInSection(kData, data.size);
+  bss.start = AllocateInSection(kBss, bss.size);
+  CHECK(text.start != nullptr && rodata.start != nullptr && data.start != nullptr &&
+        bss.start != nullptr) << "not enough space to load module on device";
+  const DevBaseAddr base_addr = low_level_device_->base_addr();
+  std::string relocated_bin = RelocateBinarySections(
+      binary_path,
+      text.start + base_addr,
+      rodata.start + base_addr,
+      data.start + base_addr,
+      bss.start + base_addr);
+  std::string text_contents = ReadSection(relocated_bin, kText);
+  std::string rodata_contents = ReadSection(relocated_bin, kRodata);
+  std::string data_contents = ReadSection(relocated_bin, kData);
+  std::string bss_contents = ReadSection(relocated_bin, kBss);
+  low_level_device_->Write(text.start, &text_contents[0], text.size);
+  low_level_device_->Write(rodata.start, &rodata_contents[0], rodata.size);
+  low_level_device_->Write(data.start, &data_contents[0], data.size);
+  low_level_device_->Write(bss.start, &bss_contents[0], bss.size);
+  SymbolMap symbol_map {relocated_bin, base_addr};
+  return BinaryInfo{
+      .text = text,
+      .rodata = rodata,
+      .data = data,
+      .bss = bss,
+      .symbol_map = symbol_map,
+  };
+}
+
+void MicroSession::SetInitBinaryPath(std::string path) {
+  init_binary_path_ = path;
+}
+
+DevAddr MicroSession::EncoderAppend(TargetDataLayoutEncoder* encoder, TVMArgs& args) {
+  auto utvm_args_slot = encoder->Alloc<UTVMArgs>();
+
+  const int* type_codes = args.type_codes;
+  int num_args = args.num_args;
+
+  auto tvm_vals_slot = encoder->Alloc<TVMValue>(num_args);
+  auto type_codes_slot = encoder->Alloc<const int>(num_args);
+
+  for (int i = 0; i < num_args; i++) {
+    switch (type_codes[i]) {
+      case kNDArrayContainer:
+      case kArrayHandle: {
+        TVMArray* arr_handle = args[i];
+        void* arr_ptr = EncoderAppend(encoder, *arr_handle).cast_to<void*>();
+        TVMValue val;
+        val.v_handle = arr_ptr;
+        tvm_vals_slot.WriteValue(val);
+        break;
+      }
+      // TODO(weberlo): Implement `double` and `int64` case.
+      default:
+        LOG(FATAL) << "Unsupported type code for writing args: " << type_codes[i];
+        break;
+    }
+  }
+  type_codes_slot.WriteRaw(type_codes, num_args);
```

Comment:
Consider add another API WriteArray, so it can handle the endianness issue.

Reply:
It seems like `WriteRaw` is already functioning as a `WriteArray` method, and it should probably be renamed to that.

To deal with endianness issues, I think we'll need to change the way we write values to the encoder.  In particular, we'll need to store the host and target endianness in the encoder and do conversions when they don't match.  Then we'll need to disallow *raw* writes and instead implement write methods for every compound and primitive data type that we want to write to the device.  This design would essentially be traits.  I think we should make this change in a future PR though, once we have more experience with different devices (e.g., RISC-V).

Reply:
TVM already provides byte swap routines in binay file handlers, I think the only effort required is to detect endianness on device, and provide the host such info.

Reply:
Let us rename to WriteArray for now. The difference is WriteRaw just performs memcpy, while WriteArray will detect endianness and do endian swapping